torch/csrc/jit/ir.cpp

```diff
@@ -413,88 +427,126 @@ void Graph::lint() const {
   // - uniques in all_nodes are unique
   // - every use will occur later in the topsort
 
-  std::unordered_set<const Value*> in_scope;
-  std::unordered_set<const Node*> node_in_scope;
-  std::unordered_set<size_t> seen_uniques;
-  std::unordered_map<const Node*, int64_t> anticipated_uses;
-  auto check_value = [&](const Value* v) {
-    auto b = in_scope.insert(v);
-    JIT_ASSERT(b.second);  // insertion took place
-    auto b2 = seen_uniques.insert(v->unique());
-    JIT_ASSERT(b2.second);  // insertion took place
-    JIT_ASSERT(v->unique() < next_unique_);
-
-    for (auto use : v->uses()) {
-      JIT_ASSERT(node_in_scope.count(use.user) == 0);
-      JIT_ASSERT(all_nodes.count(use.user) == 1);
-      anticipated_uses[use.user]++;  // int default constructs to 0
+  struct Scope {
+    Scope() {}
+    Scope(std::unique_ptr<Scope> parent)
+    : parent(std::move(parent)) {}
+    bool contains(const Value * v) {
+      return values.count(v) > 0 || (parent && parent->contains(v));
+    }
+    bool contains(const Node * n) {
+      return nodes.count(n) > 0 || (parent && parent->contains(n));
     }
+    void insert(const Value * v) {
+      JIT_ASSERT(!contains(v));
+      values.insert(v);
+    }
+    void insert(const Node * n) {
+      JIT_ASSERT(!contains(n));
+      nodes.insert(n);
+    }
+    std::unique_ptr<Scope> parent;
+  private:
+    std::unordered_set<const Value*> values;
+    std::unordered_set<const Node*> nodes;
   };
-  auto check_node = [&](const Node* n) {
-    for (auto input : n->inputs_) {
-      if (in_scope.count(input) != 1) {
-        JIT_ASSERTM(0, "%%%d not in scope", input->unique());
+  struct Env {
+    Env(const Graph & g)
+    : g(g)
+    , scope(new Scope())
+    , all_nodes_set(ALL_OF(g.all_nodes)) {} // NB: all_nodes is *unordered*
+    const Graph & g;
+    std::unique_ptr<Scope> scope;
+    std::unordered_set<size_t> seen_uniques;
+    std::unordered_map<const Node*, int64_t> anticipated_uses;
+    node_set all_nodes_set;
+    node_set sum_set;
+
+    void check_value(const Value* v) {
+      scope->insert(v);
```

Comment:
Assert that scope does not contain value before inserting?

Reply:
From the deleted code:

```
-    auto b = in_scope.insert(v);
 -    JIT_ASSERT(b.second);  // insertion took place
```

Reply:
(I briefly considered whether or not we should allow shadowing, but that would violate the "unique id uniquely identifies a memory location" invariant so it is probably best not to. I do know that sometimes maintaining "no-shadowing" can be quite a pain.)

Reply:
It's there. I just moved it into insert.
```
+    }
 +    void insert(const Node * n) {
 +      JIT_ASSERT(!contains(n));
 +      nodes.insert(n);
 +    }
```